### PR TITLE
[Form] Added an AbstractChoiceLoader

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes (deps=high failure is normal)
| Fixed tickets | -
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

Follow up #30962 and #30966,
Taking over from https://github.com/symfony/symfony/pull/30983

TODO:
- [ ] benchmark

### CODE REVIEW

File: src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+abstract class AbstractChoiceLoader implements ChoiceLoaderInterface
+{
+    /**
+     * The loaded choice list.
+     *
+     * @var ChoiceListInterface
+     */
+    protected $choiceList;
```
Comment (ro0NL) [inline src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php]: private?

**Comment by fabpot** [comment]
@yceruto What's the status of this PR? Were you able to run some benchmarks?

**Comment by yceruto** [comment]
> Were you able to run some benchmarks?

Nope :(

Status: Looking for inspiration to create some benchmarks :) I will try this week before :snowflake: fase.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/IdReader.php [hunk 2]
```diff
@@ -91,7 +91,7 @@ public function isIntId(): bool
     public function getIdValue($object)
     {
         if (!$object) {
-            return null;
+            return '';
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/IdReader.php]: This looks wrong to me. the id of `null` is not the empty string.

File: src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+abstract class AbstractChoiceLoader implements ChoiceLoaderInterface
+{
+    /**
+     * The loaded choice list.
+     *
+     * @var ChoiceListInterface
```
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php]: `ArrayChoiceList`

**Comment by HeahDude** [comment]
See #34550.

**Comment by xabbuh** [comment]
closing in favour of #34550

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/IdReader.php (hunk 2 above)
Comment (HeahDude) [reply src/Symfony/Bridge/Doctrine/Form/ChoiceList/IdReader.php]: `IdReader::getIdValue` is used as `choice_value` callback, it must always return a string actually.

**Pull request closed without merging**